zstash/globus_utils.py

```diff
@@ -0,0 +1,273 @@
+            logger.info(
+                f"Setting local_endpoint_id based on {INI_PATH}: {local_endpoint_id}"
+            )
+    else:
+        ini["local"] = {"globus_endpoint_uuid": ""}
+        try:
+            with open(INI_PATH, "w") as f:
+                ini.write(f)
+            logger.info(f"Writing to empty {INI_PATH}")
+        except Exception as e:
+            logger.error(e)
+            sys.exit(1)
+    if not local_endpoint_id:
+        fqdn = socket.getfqdn()
+        if re.fullmatch(r"n.*\.local", fqdn) and os.getenv("HOSTNAME", "NA").startswith(
+            "compy"
+        ):
+            fqdn = "compy.pnl.gov"
+        for pattern in REGEX_ENDPOINT_MAP.keys():
+            if re.fullmatch(pattern, fqdn):
+                local_endpoint_id = REGEX_ENDPOINT_MAP.get(pattern)
+                logger.info(
+                    f"Setting local_endpoint_id based on FQDN {fqdn}: {local_endpoint_id}"
+                )
+                break
+    # FQDN is not set on Perlmutter at NERSC
+    if not local_endpoint_id:
+        nersc_hostname = os.environ.get("NERSC_HOST")
+        if nersc_hostname and (
+            nersc_hostname == "perlmutter" or nersc_hostname == "unknown"
+        ):
+            local_endpoint_id = REGEX_ENDPOINT_MAP.get(r"perlmutter.*\.nersc\.gov")
+            logger.info(
+                f"Setting local_endpoint_id based on NERSC_HOST {nersc_hostname}: {local_endpoint_id}"
+            )
+    if not local_endpoint_id:
+        logger.error(
+            f"{INI_PATH} does not have the local Globus endpoint set nor could one be found in REGEX_ENDPOINT_MAP."
+        )
+        sys.exit(1)
+    return local_endpoint_id
+
+
+def get_transfer_client_with_auth(
+    both_endpoints: List[Optional[str]],
+) -> TransferClient:
+    tokens = load_tokens()
+
+    # Check if we have stored refresh tokens
+    if "transfer.api.globus.org" in tokens:
+        token_data = tokens["transfer.api.globus.org"]
+        if "refresh_token" in token_data:
+            logger.info("Found stored refresh token - using it")
+            # Create a simple auth client for the RefreshTokenAuthorizer
+            auth_client = NativeAppAuthClient(ZSTASH_CLIENT_ID)
+            transfer_authorizer = RefreshTokenAuthorizer(
+                refresh_token=token_data["refresh_token"], auth_client=auth_client
+            )
+            transfer_client = TransferClient(authorizer=transfer_authorizer)
+            return transfer_client
+
+    # No stored tokens, need to authenticate
+    logger.info("No stored tokens found - starting authentication")
+
+    # Get the required scopes
+    all_scopes = get_all_endpoint_scopes(both_endpoints)
+
+    # Use the NativeAppAuthClient pattern from the documentation
+    client = NativeAppAuthClient(ZSTASH_CLIENT_ID)
+    client.oauth2_start_flow(
+        requested_scopes=all_scopes,
+        refresh_tokens=True,  # This is the key to persistent auth!
+    )
```

Comment:
This addresses #338 -- the `oauth2_start_flow` comes from https://globus-sdk-python.readthedocs.io/en/stable/examples/native_app.html#with-refresh-tokens.

Relatedly, I believe this is the Globus Flow that @rljacob mentioned earlier today.

Reply:
I believe this `oauth2_start_flow` may also resolve the `(5) Start long transfer [...] Note that this transfer is limited to 48 hours due to Globus token expiration.` and `(6) Using Globus web interface, manually transfer zstash files that were not transferred due to token expiration.` of #339.

(Previously, that is, on `main`, we were using `native_client.login(no_local_server=True, refresh_tokens=True)`, so my hope is that the `oauth2_start_flow` is nicer to use).

But @TonyB9000 I think I'd need you to test on a long run to be sure that this is in fact the case.

Reply:
@forsyth2 This is great.  Hopefully (using the new CMIP6 production workflow) we only need to fetch an archive "as-needed", as opposed to fetching them all at once - so each fetch should last less than a day - I hope.

(I still need to understand how the last fetch (wlin's NERSC directory) failed.  I'll try it again soon.